lib/message_request.js

```diff
@@ -0,0 +1,179 @@
+    op: operation,
+    to: scope
+  };
+
+  if (!this._isValid()) {
+    throw new Error('Invalid request message.');
+  }
+};
+
+Request.buildGet = function (scope, options) {
+  return new Request('get', scope).setOptions(options);
+};
+
+Request.buildPublish = function (scope, value) {
+  var request = new Request('publish', scope);
+  request.setAttr('value', value);
+  return request;
+};
+
+Request.buildPush = function (scope, resource, action, value) {
+  var request = new Request('push', scope);
+  request.setAttr('resource', resource);
+  request.setAttr('action', action);
+  request.setAttr('value', value);
+
+  return request;
+};
+
+Request.buildNameSync = function (scope, options) {
+  return new Request('nameSync', scope).setOptions(options);
+};
+
+Request.buildSet = function (scope, value, clientData, key, userType) {
+  var request = new Request('set', scope);
+  request.setAttr('value', value);
+  request.setAttrOptionalDefined('key', key);
+  request.setAttrOptionalUndefined('type', userType);
+  if (typeof(clientData) != 'function') {
+    request.setAttrOptionalUndefined('clientData', clientData);
+  }
+  return request;
+};
+
+Request.buildSync = function (scope, options) {
+  return new Request('sync', scope).setOptions(options);
+};
+
+Request.buildSubscribe = function (scope, options) {
+  return new Request('subscribe', scope).setOptions(options);
+};
+
+Request.buildUnsubscribe = function (scope, options) {
+  return new Request('unsubscribe', scope);
+};
+
+Request.prototype.setAuthData = function (configuration) {
+  if (configuration && configuration.auth) {
+    this.setAttr('auth', configuration.auth);
+    this.setAttr('userId', configuration.userId);
+    this.setAttr('userType', configuration.userType);
+    this.setAttr('accountName', configuration.accountName);
+  }
+};
+
+// TO DO: config class should return the required data via getters
+Request.setUserData_old = function (message, configuration) {
+  message.userData = configuration && configuration.userData;
+};
+
+// TO DO: config class should return the required data via getters
+Request.setAuthData_old = function (message, configuration) {
+  if (configuration && configuration.auth) {
+    message.auth = configuration.auth;
+    message.userId = configuration.userId;
+    message.userType = configuration.userType;
+    message.accountName = configuration.accountName;
+  }
+};
+
+Request.getAttr = function (message, attr) {
+  return (message && message[attr]);
+};
+
+// Instance methods
+
+Request.prototype.getMessage = function () {
+  return this.message;
+};
+
+Request.prototype.setOptions = function (options) {
+  if (options && typeof options != 'function') {
+    this.message.options = options;
+    this.version = 2;
+  } else {
+    this.version = 1;
+  }
+
+  return this;
+};
+
+Request.prototype.isPresence = function () {
+  return this.type === 'presence';
+};
+
+Request.prototype.setAttrOptionalUndefined = function (keyName, keyValue) {
```

Comment:
let's not do this. We should not let them set attributes with keyName undefined anyway.